SILOptimizer: make vtable pruning less aggressive

compnerd opened the pull request:
A class which is marked as `internal` or `public` can be visible outside
of the current file, where the use of the VWT indirectly is possible.
If the VWT is modified and inlined, it is possible that the offsets will
no longer match resulting in an invalid dispatch.  Although this change
drastically reduces the efficacy of this pass, it ensures correctness as
a temporary measure.

Thanks to Alexander Smarus for the basis for the test case!

https://bugs.swift.org/browse/SR-13449
rdar://problem/68170038

compnerd commented:
CC: @jckarter

compnerd commented:
@swift-ci please test

gottesmm commented:
Before this goes in I would like to understand more what is happening here.

gottesmm commented:
Also note that @jckarter is on vacation right now as I understand it.

gottesmm commented on lib/SILOptimizer/Transforms/PruneVTables.cpp:
```diff
@@ -27,8 +27,13 @@ using namespace swift;
-  void runOnVTable(SILModule *M,
-                   SILVTable *vtable) {
+  void runOnVTable(SILModule *M, SILVTable *vtable) {
+    if (!vtable->getClass()->isOutermostPrivateOrFilePrivateScope()) {
```
Turning this off for everything is not appropriate, we at least should be using the tools that we already have that handle this sort of thing. There are ways of checking this visibility while handling wmo/non-wmo that we should be using rather than referring to the ast.

gottesmm replied:
Reading SILVTable, it seems that there is a query that tells you if the vtable will be serialized or was serialized. Given that we seem to not have SILLinkage here, I imagine that is what one would want to use.

@eeckstein what are your thoughts?

compnerd replied:
> Before this goes in I would like to understand more what is happening here.

This is a pretty bad regression impacting Linux and Windows builds.  Right now we incorrectly dispatch to the wrong function when the VWT is referenced outside of the module.

compnerd replied:
Can you point me to what you have in mind for checking here?  I think that the WMO doesn't guarantee that every single reference is going to get fully monomorphized right?  But, sure WMO with internal visibility should allow additional cases.

eeckstein replied:
This problem can only occur in non-WMO mode or for "open" classes in WMO mode, right?
Can we disable this optimization just for "open" classes and for non-WMO mode (which is not so important)?

slavapestov commented:
The class vtable is stored in the metadata, not the VWT.

I think a better fix would be to disable vtable pruning entirely in non-WMO mode. A more precise check would be to use vtable pruning only for fileprivate classes in non-WMO mode.

texasmichelle commented:
> Before this goes in I would like to understand more what is happening here.

Background for this PR is [SR-13449](https://bugs.swift.org/browse/SR-13449).

compnerd commented:
> The class vtable is stored in the metadata, not the VWT.

Correct.  The effective VWT in the metadata is what is being pruned.

> I think a better fix would be to disable vtable pruning entirely in non-WMO mode. A more precise check would be to use vtable pruning only for fileprivate classes in non-WMO mode.

Yes, I agree with you, and that is what I would prefer to do:
- fileprivate/private only classes always
- in WMO, add internal classes as well

I suppose that we could disable vtable pruning entirely in non-WMO mode, but if we can do it safely, why not?

gottesmm commented:
@slavapestov question. I imagine the issue here in the design space is that we never have given vtables a SILLinkage and the infrastructure around isExternallyUsedSymbol. Your thoughts?

My thought is given that this causes a real crash, I think it makes sense to just disable this in this way for now. Really we should just fix the vtable to have VTables.

compnerd commented on test/SILOptimizer/prune-vtables-cross-file.swift:
```diff
@@ -0,0 +1,9 @@
+// RUN: %target-swift-frontend -module-name M -Osize %S/Inputs/cross-file/main.swift %S/Inputs/cross-file/support.swift -S | %FileCheck %s
+
+// REQUIRE: OS=linux-gnu
```
Bleh, this should be `REQUIRES`.

compnerd committed:
SILOptimizer: make vtable pruning less aggressive

A class which is marked as internal or public can be visible outside
of the current file, where the use of the VWT indirectly is possible.
If the VWT is modified and inlined, it is possible that the offsets will
no longer match resulting in an invalid dispatch.  Limit the pass to
when WMO is enabled or the type is private in non-WMO cases.

compnerd commented:
@slavapestov, @eeckstein - okay, the pass always runs under WMO now and only scans private classes on non-WMO mode.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

slavapestov commented on lib/SILOptimizer/Transforms/PruneVTables.cpp:
```diff
@@ -27,10 +27,16 @@ using namespace swift;
     LLVM_DEBUG(llvm::dbgs() << "PruneVTables inspecting table:\n";
                vtable->print(llvm::dbgs()));
+    if (!M->isWholeModule() &&
+        !vtable->getClass()->isOutermostPrivateOrFilePrivateScope()) {
```
I would prefer `getClass()->getEffectiveLinkage() < AccessLevel::Internal`

slavapestov replied:
Sorry, getEffectiveAccess()

compnerd replied:
I like that better as well.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

eeckstein reviewed (approved):
lgtm!

slavapestov commented on lib/SILOptimizer/Transforms/PruneVTables.cpp:
```diff
@@ -27,10 +27,16 @@ using namespace swift;
     LLVM_DEBUG(llvm::dbgs() << "PruneVTables inspecting table:\n";
                vtable->print(llvm::dbgs()));
+    if (!M->isWholeModule() &&
+        vtable->getClass()->getEffectiveAccess() >= AccessLevel::FilePrivate) {
```
Shouldn't it be `>` not `>=` then?

compnerd replied:
Oops, I thought Id fixed that.  #33808

jckarter replied:
I think this change is unnecessarily pessimistic. The check for `calleesAreStaticallyKnowable ` [below](https://github.com/apple/swift/pull/33770/files#diff-fa3907ca639ee730d10b9c0a357f10d2L76) should already prevent us from pruning nonfinal internal vtable entries. We should still be able to prune entries for final methods as long as the final method is not itself an override, because it should never need vtable dispatch to begin with.

jckarter replied:
It looks like if a method is originally final we already don't give it a vtable entry at all. So, we should be able to remove the `if (!methodDecl->isFinal())` precondition below and get the same effect. I think that's better than reimplementing `calleesAreStaticallyKnowable` here.
